packages/vite/src/node/plugins/importAnalysisBuild.ts

```diff
@@ -73,30 +72,97 @@ function preload(
+        const currentPath = possibleLinks[i].href
+        if (
+          currentPath === dep ||
+          new URL(currentPath, importerUrl).href === dep
+        ) {
+          return
+        }
       }
-      // @ts-ignore
-      const link = document.createElement('link')
-      // @ts-ignore
-      link.rel = isCss ? 'stylesheet' : scriptRel
-      if (!isCss) {
-        link.as = 'script'
-        link.crossOrigin = ''
+
+      const preloadInImg = (res: () => void) => {
+        const img = new Image()
+        img.onerror = () => res()
+        img.onload = () => res()
+        img.src = dep
+      }
+
+      const loadLink = () => {
+        // @ts-ignore
+        const link = document.createElement('link')
+        // @ts-ignore
+        link.rel = isCss ? 'stylesheet' : scriptRel
+        if (!isCss) {
+          link.as = 'script'
+          link.crossOrigin = ''
+        }
+        link.href = dep
+        // @ts-ignore
+        document.head.appendChild(link)
+        return link
+      }
+
+      const waitForLoad = (link: HTMLLinkElement) => {
+        // @ts-ignore
+        if (__VITE_IS_MODERN__) {
+          return new Promise<void>((res, rej) => {
+            link.addEventListener('load', () => res())
+            link.addEventListener('error', () =>
+              rej(new Error(`Unable to preload CSS for ${dep}`))
+            )
+          })
+        } else {
+          return Promise.all<void>([
+            new Promise((res, rej) => {
+              // We query the path (that should be cached already), for knowing if there is an error while downloading it. (this is the only reason)
+              const req = new XMLHttpRequest()
+              req.addEventListener('load', () => res())
+              req.addEventListener('error', () =>
+                rej(
+                  new Error(
+                    `Unable to preload CSS (via XMLHttpRequest) for ${dep}`
+                  )
+                )
+              )
+              req.open('GET', dep)
+              req.send()
+            }),
+            new Promise((res) => {
+              // On legacy browsers, let them a chance to process the newly referred CSS link.
+              setTimeout(res)
+            })
+          ]) as unknown as Promise<void>
+        }
       }
-      link.href = dep
+
+      const loadLinkAndWait = () => {
+        return waitForLoad(loadLink())
+      }
+
       // @ts-ignore
-      document.head.appendChild(link)
-      if (isCss) {
-        return new Promise((res, rej) => {
-          link.addEventListener('load', res)
-          link.addEventListener('error', () =>
-            rej(new Error(`Unable to preload CSS for ${dep}`))
-          )
-        })
+      if (__VITE_IS_MODERN__) {
+        if (isCss) {
+          return loadLinkAndWait()
+        } else {
+          loadLink()
+        }
+      } else {
+        if (isCss) {
+          return new Promise<void>(preloadInImg).then(loadLinkAndWait)
```

Comment:
> Maybe we could simplify this to `XHR -> add link` from `load img -> add link -> XHR`?

I'm not sure about browsers behavior.
There's some feeling in my stomach that if you load a content as image, it is "much more cached" in some sense (e.g. having two images with the same source in the same page, will probably look always the same to each other, even if their source isn't static).

As you can see in https://github.com/vitejs/vite/blob/e72c0905bec8b0540fa41622956d26028beed04e/packages/vite/src/node/plugins/importAnalysisBuild.ts#L127-L138, the only reason I'm calling XHR **is just for the ability to know when there is an error on loading the content**.
Shell we just don't care about reporting error on loading in legacy? This will simplify the flow to Img -> Style

Reply:
By "much more cached" - I mean that it seems like there is "cache" and there is "page resources".
When loading to image, it's get loaded to both of them, and on XHR, it seems to be loaded only to the cache.

If we load it to the page resources, it feels like the CSS will be immediately available after, unlike restoring it from the cache.

Reply:
See now the new preload function, I had stripped off the XHR request, resulting in removing the ability to detect an error on fetching the CSS data when on legacy (I don't think it's so important, and makes things smoother):
https://github.com/vitejs/vite/blob/6dbe67656040aab79e9c9c0e32944f56c0856ea2/packages/vite/src/node/plugins/importAnalysisBuild.ts#L58-L147